File: dev_ws/src/learning_service_python/learning_service_python/person_info_client.py
"""
 @Description: 
 @version: v1.0
 @Author: HTY
 @Date: 2022-07-22 17:17:59
"""

from learning_interface.srv import PersonInfo
import sys
import rclpy
from rclpy.node import Node


class PersonInfoClient(Node):
    def __init__(self, node):
        super(PersonInfoClient, self).__init__(node)
        self.client = self.create_client(PersonInfo, "person_info_service")
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Service not available, waiting...")
        self.req = PersonInfo.Request()
        self.future = None

    def send_person_info(self):
        self.req.name = str(sys.argv[1])
        self.future = self.client.call_async(self.req)


def main(args=None):
    rclpy.init(args=args)

    person_info_client = PersonInfoClient("person_info_client_node")
    person_info_client.send_person_info()

    while rclpy.ok():
        rclpy.spin_once(person_info_client)
        if person_info_client.future.done():
            try:
                response = person_info_client.future.result()
            except Exception as e:
                person_info_client.get_logger().warn("Service call failed: %s" % e)
            else:
                person_info_client.get_logger().info(f"{response.message}")
            break

    person_info_client.destroy_node()
    rclpy.shutdown()

        
if __name__ == "__main__":
    main()

File: dev_ws/src/learning_service_cpp/src/person_info_client.cpp
/*
 * @Description: ROS2 service客户端 发送自定义消息 cpp版本
 * @version: v1.0
 * @Author: HTY
 * @Date: 2022-07-22 20:38:40
 */

#include "rclcpp/rclcpp.hpp"
#include "learning_interface/srv/person_info.hpp"


class PersonInfoClient : public rclcpp::Node
{
private:
    rclcpp::Client<learning_interface::srv::PersonInfo>::SharedPtr client;
    void person_info_callback(rclcpp::Client<learning_interface::srv::PersonInfo>::SharedFuture response)
    {
        const auto & result = response.get();
        RCLCPP_INFO_STREAM(this->get_logger(), result->message << std::endl <<
        "age: " << (int) result->age << " id: " << result->id);
    }

public:
    explicit PersonInfoClient(const std::string & node) : rclcpp::Node(node)
    {
        client = this->create_client<learning_interface::srv::PersonInfo>("person_info_service");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_INFO_STREAM(this->get_logger(), "Waiting for service started...");
        }
    }

    void send_person_info(int argc, char* argv[])
    {
        auto request = std::make_shared<learning_interface::srv::PersonInfo::Request>();
        request->name = argv[1];
        client->async_send_request(request, std::bind(&PersonInfoClient::person_info_callback, this, std::placeholders::_1));
    }
};


int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto person_info_client = std::make_shared<PersonInfoClient>("person_info_client_node");
    person_info_client->send_person_info(argc, argv);
    rclcpp::spin(person_info_client);
    rclcpp::shutdown();
    return 0;
}